src/SortingAlgorithms.jl

```diff
@@ -631,4 +683,308 @@ else
+    @assert lo < m < hi
+    a = lo
+    b = m + 1
+    lenA = 1 + m - lo
+    lenB = hi - m
+
+    # regular merge if buffer is big enough
+    if lenA <= length(buf)
+        merge!(v, lo, m, hi, o, buf)
+        return
+    elseif lenB <= length(buf)
+        # TODO ?
+        # does not occur in balanced mergesort where length(A) <= length(B)
+        error("not implemented")
+        return
+    end
+
+    len = lenA + lenB
+    pagesize = isqrt(len)
+    nPages = len ÷ pagesize # a partial page at the end does not count
+    @assert length(buf) >= 3pagesize
+    @assert length(pageLocations) >= nPages - 3
+
+    @inline page_offset(page) = (page-1)*pagesize + lo - 1
+
+    @inbounds begin
+        ##################
+        # merge
+        ##################
+        # merge the first 3 pages into buf
+        a,b,_ = merge!((_,_,k) -> k<=3pagesize,buf,v,v,o,a,b,1)
+        # initialize variables for merging into pages
+        pages = Pages(-17, 0, 1, (m-lo) ÷ pagesize + 2) # first argument is unused
+        # more efficient loop while more than pagesize elements of A and B are remaining
+        while_condition1(offset) = (_,_,k) -> k <= offset + pagesize
+        while a < m-pagesize && b < hi-pagesize
+            pages = next_page!(pageLocations, pages, pagesize, lo, a)
+            offset = page_offset(pages.current)
+            a,b,_ = merge!(while_condition1(offset),v,v,v,o,a,b,offset+1)
+        end
+        # merge until either A or B is empty or the last page is reached
+        k, offset = nothing, nothing
+        while_condition2(offset) = (a,b,k) -> k <= offset + pagesize && a <= m && b <= hi
+        while a <= m && b <= hi && pages.currentNumber + 3 < nPages
+            pages = next_page!(pageLocations, pages, pagesize, lo, a)
+            offset = page_offset(pages.current)
+            a,b,k = merge!(while_condition2(offset),v,v,v,o,a,b,offset+1)
+        end
+        # if the last page is reached, merge the remaining elements into the final partial page
+        if pages.currentNumber + 3 == nPages && a <= m && b <= hi
+            a,b,k = merge!((a,b,_) -> a <= m && b <= hi,v,v,v,o,a,b,nPages*pagesize+lo)
+            copyto!(v, k, v, a <= m ? a : b, hi-k+1)
+        else
+            use_a = a <= m
+            # copy the incomplete page
+            partial_page_size = offset + pagesize - k + 1
+            copyto!(v, k, v, use_a ? a : b, partial_page_size)
+            use_a && (a += partial_page_size)
+            use_a || (b += partial_page_size)
+            # copy the remaining full pages
+            while use_a ? a <= m - pagesize + 1 : b <= hi - pagesize + 1
+                pages = next_page!(pageLocations, pages, pagesize, lo, a)
+                offset = page_offset(pages.current)
+                copyto!(v, offset+1, v, use_a ? a : b, pagesize)
+                use_a && (a += pagesize)
+                use_a || (b += pagesize)
```

Comment:
This is funny and a little off topic: I prefer each style in this list over the one immediately before it
```julia
use_a && (a += pagesize)
use_a || (b += pagesize)
```
```julia
use_a ? (a += pagesize) : (b += pagesize)
```
```julia
if use_a (a += pagesize) else (b += pagesize) end
```
```julia
if use_a
    a += pagesize
else 
    b += pagesize
end
```
```julia
use_a && (a += pagesize)
use_a || (b += pagesize)
```
(yes, that is a circular preference loop, it's probably fine as is)

Reply:
I agree.